### PR TITLE
Set fields as readonly

### PR DESCRIPTION
**Pull request by amanmahajan7**

### CODE REVIEW

File: src/utils/selectedCellUtils.ts [hunk 1]
```diff
@@ -96,7 +99,6 @@ export function getNextSelectedCellPosition<R, SR>({
           idx: 0
         };
       }
-      setColSpan(true);
```
Comment (amanmahajan7) [inline src/utils/selectedCellUtils.ts:99]: It is not required as the next position is going to be the first cell of a colSpan range

File: src/hooks/useViewportRows.ts [hunk 2]
```diff
@@ -1,6 +1,15 @@
...
+import type { GroupRow, RowHeightArgs } from '../types';
+
+type GroupByDictionary<TRow> = Record<
+  string,
+  {
+    childRows: readonly TRow[];
+    childGroups: readonly TRow[] | Readonly<GroupByDictionary<TRow>>;
+    startRowIndex: number;
+  }
+>;
```
Comment (nstepien) [inline src/hooks/useViewportRows.ts:12]: Should work,, no?
```suggestion
type GroupByDictionary<TRow> = Readonly<Record<
  string,
  {
    readonly childRows: readonly TRow[];
    readonly childGroups: readonly TRow[] | GroupByDictionary<TRow>;
    readonly startRowIndex: number;
  }
>>;
```

File: src/hooks/useViewportRows.ts [hunk 3]
```diff
@@ -38,7 +47,7 @@ export function useViewportRows<R>({
       rows: readonly R[],
       [groupByKey, ...remainingGroupByKeys]: readonly string[],
       startRowIndex: number
-    ): [GroupByDictionary<R>, number] => {
+    ): [Readonly<GroupByDictionary<R>>, number] => {
```
Comment (nstepien) [inline src/hooks/useViewportRows.ts:50]:
```suggestion
    ): [GroupByDictionary<R>, number] => {
```

File: src/types.ts [hunk 4]
```diff
@@ -175,7 +170,7 @@ export interface RowRendererProps<TRow, TSummaryRow = unknown>
 }
 
 export interface RowsChangeData<R, SR = unknown> {
-  indexes: number[];
+  indexes: readonly number[];
```
Comment (nstepien) [inline src/types.ts:173]: We generate it then pass it on to the users, it's fine to leave it mutable, users can manipulate if they want to.
```suggestion
  indexes: number[];
```

**Review by nstepien — CHANGES_REQUESTED** [review]

File: src/hooks/useViewportRows.ts (hunk 2 above)
Comment (amanmahajan7) [reply src/hooks/useViewportRows.ts:12]: We [mutate](https://github.com/adazzle/react-data-grid/blob/472d2d27444b1251a59623005ac6b686c0fdb8c5/src/hooks/useViewportRows.ts#L59) the value but I can set fields to readonly

File: src/hooks/useViewportRows.ts [hunk 5]
```diff
@@ -1,6 +1,15 @@
 import { useMemo } from 'react';
 import { floor, max, min } from '../utils';
-import type { GroupRow, GroupByDictionary, RowHeightArgs } from '../types';
+import type { GroupRow, RowHeightArgs } from '../types';
+
+type GroupByDictionary<TRow> = Record<
+  string,
+  {
+    readonly childRows: readonly TRow[];
+    readonly childGroups: readonly TRow[] | GroupByDictionary<TRow>;
```
Comment (nstepien) [inline src/hooks/useViewportRows.ts:9]:
```suggestion
    readonly childGroups: readonly TRow[] | Readonly<GroupByDictionary<TRow>>;
```

File: src/utils/selectedCellUtils.ts [hunk 6]
```diff
@@ -64,19 +63,23 @@ export function getNextSelectedCellPosition<R, SR>({
         if (colIdx > posIdx) break;
         const colSpan = getColSpan(column, lastFrozenColumnIndex, { type: 'ROW', row });
         if (colSpan && posIdx > colIdx && posIdx < colSpan + colIdx) {
-          position.idx = colIdx + (moveRight ? colSpan : 0);
-          break;
+          return {
+            ...position,
```
Comment (nstepien) [inline src/utils/selectedCellUtils.ts:67]: It's just 1 property, so we can set it manually and avoid the iterating over the object by spreading it.
```suggestion
            rowIdx: position.rowIdx,
```

File: src/utils/selectedCellUtils.ts [hunk 7]
```diff
@@ -51,9 +51,8 @@ export function getNextSelectedCellPosition<R, SR>({
   isGroupRow
 }: GetNextSelectedCellPositionOpts<R, SR>): Position {
   const rowsCount = rows.length;
-  let position = nextPosition;
 
-  const setColSpan = (moveRight: boolean) => {
+  const setColSpan = (position: Position, moveRight: boolean): Position => {
```
Comment (nstepien) [inline src/utils/selectedCellUtils.ts:55]: Does the name `setColSpan` make sense anymore?

I wonder if we even need a position object? Could we have something like this?
```ts
let { idx, rowIdx } = nextPosition;
```
Then we update `idx`/`rowIdx`, and at the end
```ts
return { idx, rowIdx };
```

Comment (amanmahajan7) [reply src/utils/selectedCellUtils.ts:55]: I will check

**Review by nstepien — APPROVED** [review]

**Pull request merged**